lib/src/index.rs

```diff
@@ -28,49 +29,73 @@ use crate::operation::Operation;
 use crate::store::Store;
 
-/// Error while reading index from the `IndexStore`.
+/// Returned if an error occurs while reading an index from the [`IndexStore`].
 #[derive(Debug, Error)]
 #[error(transparent)]
 pub struct IndexReadError(pub Box<dyn std::error::Error + Send + Sync>);
 
-/// Error while writing index to the `IndexStore`.
+/// Returned if an error occurs while writing an index to the [`IndexStore`].
 #[derive(Debug, Error)]
 #[error(transparent)]
 pub struct IndexWriteError(pub Box<dyn std::error::Error + Send + Sync>);
 
-/// Error to be returned if `Index::all_heads_for_gc()` is not supported by the
+/// An error returned if `Index::all_heads_for_gc()` is not supported by the
 /// index backend.
 #[derive(Debug, Error)]
 #[error("Cannot collect all heads by index of this type")]
 pub struct AllHeadsForGcUnsupported;
 
+/// Defines the interface for types that provide persistent storage for an
+/// index.
 pub trait IndexStore: Send + Sync + Debug {
+    #[allow(missing_docs)]
     fn as_any(&self) -> &dyn Any;
 
+    /// Returns a name representing the type of index that the `IndexStore` is
+    /// compatible with. For example, the `IndexStore` for the default index
+    /// returns "default".
     fn name(&self) -> &str;
 
+    /// Returns the index at the specified operation.
     fn get_index_at_op(
         &self,
         op: &Operation,
         store: &Arc<Store>,
     ) -> Result<Box<dyn ReadonlyIndex>, IndexReadError>;
 
+    /// Writes `index` to the index store and returns a read-only version of the
+    /// index.
     fn write_index(
         &self,
         index: Box<dyn MutableIndex>,
         op_id: &OperationId,
     ) -> Result<Box<dyn ReadonlyIndex>, IndexWriteError>;
 }
 
+/// Defines the interface for types that provide an index of the commits in a
+/// repository by [`CommitId`].
 pub trait Index: Send + Sync {
+    /// Returns the minimum prefix length to disambiguate `commit_id` from other
+    /// commits in the index. The length returned is the number of hexadecimal
+    /// digits in the minimum prefix.
+    ///
+    /// If the given `commit_id` doesn't exist, zero is returned since a prefix
+    /// length of zero never matches any commit id.
     fn shortest_unique_commit_id_prefix_len(&self, commit_id: &CommitId) -> usize;
 
+    /// Searches the index for commit IDs matching `prefix`. Returns a
+    /// [`PrefixResolution`] with a [`CommitId`] if the prefix matches a single
+    /// commit.
     fn resolve_commit_id_prefix(&self, prefix: &HexPrefix) -> PrefixResolution<CommitId>;
 
+    /// Returns true if `commit_id` is present in the index.
     fn has_id(&self, commit_id: &CommitId) -> bool;
 
+    /// Returns true if `ancestor_id` commit is an ancestor of the
+    /// `descendant_id` commit.
     fn is_ancestor(&self, ancestor_id: &CommitId, descendant_id: &CommitId) -> bool;
 
+    /// Returns the common ancestors of the commits in `set1` and `set2`.
```

Comment:
I updated the comment, but I'm not sure if I understand. Wouldn't the greatest common ancestor be the root commit in all cases?

What would this function return if we fed it a set1 with `e1b91b95` and set2 with `fb24d819`?

```
@  rrzzxslo e1b91b95
│  (empty) four
│ ◉  qsosxrlq fb24d819
├─╯  (empty) three
◉  ntlprspw 403096b0
│  (empty) two
◉  wyktpyvy  f7103dfb
│  (empty) one
◉  zzzzzzzz root() 00000000
```

Reply:
It would only return `403096b0 ` (the "child-most" common ancestor). Maybe "closest" is a better adjective.

Reply:
Ok got it. I would say closest or nearest ancestor. "Child-most" sounded to me like the ancestor with the most descendants, or something like that.

But if we are returning the closest common ancestor, when are multiple commits returned? Does that only happen if a merge commit is duplicated? Maybe there is another scenario where it occurs that I'm not thinking of?

Reply:
> But if we are returning the closest common ancestor, when are multiple commits returned? Does that only happen if a merge commit is duplicated? Maybe there is another scenario where it occurs that I'm not thinking of?

It's a pretty non-obvious case: so-called "criss-cross merges". See https://tonyg.github.io/revctrl.org/CrissCrossMerge.html

Reply:
Oh yes, I've seen this before in an example somewhere. I took a look at the tests though and I don't think that's the only time there are multiple commits in the output.

Actually, I also realized that the comment I added is imprecise, since it doesn't explain the difference between `set1` and `set2`. If we simply returned the common ancestor among the commits then we'd only need one set.

From the tests it looks like perhaps the output set is the result of finding the nearest ancestor of the cross product of the commits in the two sets? And to be precise, the nearest ancestor is the ancestor that requires traversing the fewest edges. And, like in the other functions, a commit is also an ancestor of itself.

Take this example:
https://github.com/martinvonz/jj/blob/d661f59f9d65bc0aa47e19fc9436e51bc884873b/lib/src/default_index/mod.rs#L1014-L1019
https://github.com/martinvonz/jj/blob/d661f59f9d65bc0aa47e19fc9436e51bc884873b/lib/src/default_index/mod.rs#L1071-L1074

The nearest ancestor of `id_1` and `id_5` is 1, which requires traversing 2 edges (as opposed to 3 for `id_0`). And the nearest ancestor of `id_2` and `id_5` is `id_2`.

But then if I add a new commit `id_6` with parent `id_4` and have `set_1 = &[id_5.clone(), id_6.clone()]` and `set2 = &[id_2.clone()]`, the result is `[id_2]`, which isn't what we'd expect from the cross product of the sets since the nearest common ancestor of `id_6` and `id_2` is `id_0`.

That seems wrong? Or at least not internally consistent?  If the correct description of the function is that it returns the nearest ancestor among all ancestors in `set1` and `set2`, then I'd expect the (1,2),(5) test case to return (2), instead of (1,2).

Maybe I need to think about this more, but I'm not sure I understand it completely at this point. I pushed another commit with the new test case I mentioned.

Reply:
I stepped over the code. The reason `common_ancestors` for (1,2) and (5) = (1,2) is that 5 has two ancestors, 4 and 2. Both are added second set for consideration here:

https://github.com/martinvonz/jj/blob/6626c93d3843fb0804c3e6be49b22b26834f1fc8/lib/src/default_index/composite.rs#L310-L312

Since the other set contains (1, 2), both sets now have two entries and eventually the result is (1, 2). It progresses like this:

- (1,2) and (5)
- (1,2) and (2,4)
- (1,2) and (1, 2)
- (1) and (1) with output (2)
- () and () with output (1,2)

For (5,6) + (2) = (2), the right hand side never has more than one entry, so only 2 is returned.

- (5,6) and (2)
- (4, 5) and (2)
- (4, 2) and (2)
- (4) and () with output (2)

I guess the part that seemed odd to me is that while 1 and 2 are both ancestors of 5, they are not equidistant to 5. Only one edge separates 2 and 5, but two edges separate 1 and 5. Do we care about this, or is the current implementation ok?

If the current behavior is ok, then maybe the proposed comment (below) is fine? I'm not sure how to make it more precise.

```
    /// Returns the nearest common ancestor or ancestors of the commits in
    /// `set1` and `set2`.
```

@yuja @martinvonz Thank you for your patience and all of the feedback you gave. I think this is the last open item for this PR. If we can make a decision here then we can close this out.

Reply:
I think the way to think of it is to imagine a commit created with all the commits in set 1 as input and another commit created with all the commits in set 2 as input, and then you find the nearest common ancestors between those. I've seen that explanation somewhere in git's or hg's documentation, I think. I hope that clarifies. Thanks for asking.

Reply:
I guess the analogous function for git is `git merge-base`? 

- https://git-scm.com/docs/git-merge-base

The thing is, I re-created the examples above with git, and for the (1,2) and (5) case, `git merge-base` only outputs commit (2), not (1,2).

The git merge-base docs also seem to suggest there will only be multiple best common ancestors if there's a criss-cross merge. The (1,2) and (5) case isn't a criss-cross merge.

> When the history involves criss-cross merges, there can be more than one best common ancestor for two commits.

I'm just having trouble convincing myself that this function does that it says on the tin. In the (5,6) and (2) example, (2) is not an ancestor of 6, but it is the best merge base of (5,6) and (2). But then when we look at the (1,2) and (5) example, the output (1,2) is not the best merge base of the inputs, but all of the inputs a common ancestor in the output.

Doesn't this seem inconsistent? In one case one of the inputs (6) doesn't have an ancestor in the output, but we return the best merge base. In the other case all of the inputs have an ancestor in the output, but we don't return the best merge base.

Reply:
> The thing is, I re-created the examples above with git, and for the (1,2) and (5) case, `git merge-base` only outputs commit (2), not (1,2).

Good point! I suppose I didn't think enough about the case where there are several ancestors of one set in the other set. OTOH, I'm not sure why we need to keep more than one arbitrary nearest ancestor. I think we only use it when merging commits, where we recursively merge common ancestors. That's a trick I copied from Git. See ttps://git-scm.com/docs/merge-strategies#Documentation/merge-strategies.txt-recursive. I'm not sure when it actually matters. I won't have time to look into it either, unfortunately (I'm curious about it).

I think by "nearest", we just mean that there's no descendant of that commit that is also a common ancestor. I don't think it needs to influence the choice between commits 1 and 2.

Reply:
> I think by "nearest", we just mean that there's no descendant of that commit that is also a common ancestor.

Yeah, I think that's right. Git uses the word "best" here, and I think nearest confused me a little bit because there's no real definition of nearest that makes 1 equivalent to 2 in the example above.

I checked again and if you pass `--all` to `get merge-base` in the (1,2) and (5) example it outputs both commits. It also says:

> When the --all option is not given, it is unspecified which best one is output.

So maybe this is all a misunderstanding on my part about what "nearest" means? Anyways, I replaced "nearest" with "best' and described what best means in the updated documentation. PTAL and LGTM if everything looks good so that we can close this out.

Thanks!!!